### PR TITLE
Add school cohorts step 2: Use school_cohort everywhere

### PR DESCRIPTION
**Pull request by kicferk1**
### Context
https://dfedigital.atlassian.net/browse/CPDRP-545

We currently duplicate `school_cohort` and `school` and `cohort` data on participant profiles.
This makes everything use `school_cohort` in all the places.

I will remove `cohort` in next PR.

### Changes proposed in this pull request
1. Remove school and cohort associations from ect and mentor participant profiles
2. Fix everything that's broken by it

On Monday I might split it into more commits so it's easier to read through.

### CODE REVIEW

File: app/controllers/admin/participants_controller.rb [hunk 1]
```diff
@@ -17,7 +17,7 @@ def index
   private
 
     def load_participant
-      @participant = User.is_participant.find(params[:id])
+      @participant = User.is_ecf_participant.find(params[:id])
```
Comment (ethanmills) [inline app/controllers/admin/participants_controller.rb:20]: Why this change? This should be all participants I think?

Comment (ethanmills) [reply app/controllers/admin/participants_controller.rb:20]: NPQ will use this interface for manual validation

File: app/controllers/admin/schools/participants_controller.rb [hunk 2]
```diff
@@ -7,12 +7,7 @@ class Schools::ParticipantsController < Admin::BaseController
     before_action :set_school
 
     def index
-      @participants = User.where(id: @school.participant_profiles
-                                            .active
-                                            .ecf
-                                            .select(:user_id))
-                          .order(:full_name)
-                          .includes(participant_profile: %i[cohort school])
+      @participants = @school.participants.order(:full_name)
```
Comment (ethanmills) [inline app/controllers/admin/schools/participants_controller.rb:10]: Missing includes?

Comment (ethanmills) [reply app/controllers/admin/schools/participants_controller.rb:10]: This shows withdrawn participants, right?

File: app/controllers/schools/add_participants_controller.rb [hunk 3]
```diff
@@ -11,8 +11,7 @@ class AddParticipantsController < ::Schools::BaseController
     result as: :participant_profile
 
     abandon_journey_path do
-      participants = User.is_participant.in_school(@school.id)
-      participants.any? ? schools_participants_path : schools_cohort_path
+      @school_cohort.participants.any? ? schools_participants_path : schools_cohort_path
```
Comment (ethanmills) [inline app/controllers/schools/add_participants_controller.rb:14]: Does this include withdrawn participants?

File: app/forms/schools/add_participant_form.rb [hunk 4]
```diff
@@ -65,13 +65,11 @@ def type_options
     end
 
     def can_add_self?
-      User.is_participant.in_school(school_cohort.school).exclude? current_user
+      school_cohort.participants.exclude? current_user
```
Comment (ethanmills) [inline app/forms/schools/add_participant_form.rb:68]: Not sure how this plays with withdrawn profiles

File: app/controllers/admin/participants_controller.rb (hunk 1 above)
Comment (kicferk1) [reply app/controllers/admin/participants_controller.rb:22]: oops

File: app/controllers/admin/schools/participants_controller.rb (hunk 2 above)
Comment (kicferk1) [reply app/controllers/admin/schools/participants_controller.rb:10]: Missing? And what would you like to include in there? :P

a school? a school cohort? a cohort?

The technical decision of including NPQ profiles in ParticipantProfiles, coupled with those profiles looking nothing like ecf profiles means I can't include anything meaningful here...

Comment (ethanmills) [reply app/controllers/admin/schools/participants_controller.rb:10]:
```
includes(
      early_career_teacher_profile: %i[cohort school],
      mentor_profile: %i[cohort school],
)
```

File: app/controllers/admin/participants_controller.rb [hunk 5]
```diff
@@ -10,8 +10,10 @@ class ParticipantsController < Admin::BaseController
     def show; end
 
     def index
-      @participant_profiles = policy_scope(ParticipantProfile)
-        .ransack(user_full_name_or_school_name_or_school_urn_cont: params[:query]).result
```
Comment (ethanmills) [inline app/controllers/admin/participants_controller.rb:14]: Why was this change needed?

File: app/controllers/admin/schools/participants_controller.rb [hunk 6]
```diff
@@ -7,12 +7,10 @@ class Schools::ParticipantsController < Admin::BaseController
...
-      @participants = User.where(id: @school.participant_profiles
-                                            .active
-                                            .ecf
-                                            .select(:user_id))
-                          .order(:full_name)
-                          .includes(participant_profile: %i[cohort school])
+      @participants = @school.active_participants.order(:full_name).includes(
+        early_career_teacher_profile: [:mentor, school_cohort: %i[school cohort]],
+        mentor_profile: [school_cohort: %i[school cohort]],
+      )
```
Comment (ethanmills) [inline app/controllers/admin/schools/participants_controller.rb:13]: This is going to include NPQ ppts, we've lost the `ecf` scope.

File: app/controllers/schools/add_participants_controller.rb [hunk 7]
```diff
@@ -11,8 +11,7 @@ class AddParticipantsController < ::Schools::BaseController
     result as: :participant_profile
 
     abandon_journey_path do
-      participants = User.is_participant.in_school(@school.id)
-      participants.any? ? schools_participants_path : schools_cohort_path
+      @school_cohort.active_participants.any? ? schools_participants_path : schools_cohort_path
```
Comment (ethanmills) [inline app/controllers/schools/add_participants_controller.rb:14]: This will break if they have an NPQ ppt

File: app/forms/participant_mentor_form.rb [hunk 8]
```diff
@@ -13,9 +13,7 @@ def mentor
   end
 
   def available_mentors
-    User.where(
-      id: school.mentor_profiles_for(Cohort.find(cohort_id)).select(:user_id),
-    ).order(:full_name)
+    SchoolCohort.find_by(school_id: school_id, cohort_id: cohort_id).mentors.order(:full_name)
```
Comment (ethanmills) [inline app/forms/participant_mentor_form.rb:16]: Will this include withdrawn mentors?

File: app/forms/schools/add_participant_form.rb [hunk 9]
```diff
@@ -65,13 +65,11 @@ def type_options
...
     def can_add_self?
-      User.is_participant.in_school(school_cohort.school).exclude? current_user
+      school_cohort.active_participants.exclude? current_user
     end
 
     def mentor_options
-      @mentor_options ||= User.where(
-        id: school_cohort.school.mentor_profiles_for(school_cohort.cohort).select(:user_id),
-      ).order(:full_name)
+      @mentor_options ||= school_cohort.mentors.order(:full_name)
```
Comment (ethanmills) [inline app/forms/schools/add_participant_form.rb:72]: This includes withdrawn mentors

File: app/models/participant_profile/npq.rb [hunk 10]
```diff
@@ -1,6 +1,7 @@
 # frozen_string_literal: true
 
 class ParticipantProfile::NPQ < ParticipantProfile
+  self.ignored_columns = %i[mentor_profile_id]
```
Comment (ethanmills) [inline app/models/participant_profile/npq.rb:4]: school_cohort_id?

File: app/models/user.rb [hunk 11]
```diff
@@ -6,6 +6,8 @@ class User < ApplicationRecord
 
   has_one :induction_coordinator_profile, dependent: :destroy
   has_many :schools, through: :induction_coordinator_profile
+  has_many :managed_participant_profiles, through: :schools, source: :participant_profiles
```
Comment (ethanmills) [inline app/models/user.rb:9]: What does this mean?

File: app/models/user.rb [hunk 12]
```diff
@@ -94,20 +96,12 @@ def school
...
   }
 
-  scope :includes_school, lambda {
-    includes(
-      early_career_teacher_profile: %i[cohort school],
-      mentor_profile: %i[cohort school],
-      npq_profiles: :school,
-    )
+  scope :is_ecf_participant, lambda {
+    includes(participant_profile: :school_cohort).joins(:participant_profile).merge(ParticipantProfile.ecf.active)
```
Comment (ethanmills) [inline app/models/user.rb:100]: It scares me that `has_one :participant_profile` not `has_many :participant_profiles`

File: app/policies/participant_profile_policy.rb [hunk 13]
```diff
@@ -6,7 +6,7 @@ def resolve
       return scope.all if user.admin?
       return scope.none unless user.induction_coordinator?
 
-      scope.where(school_id: user.induction_coordinator_profile.schools.select(:id))
+      user.managed_participant_profiles
```
Comment (ethanmills) [inline app/policies/participant_profile_policy.rb:9]: This is not the same thing

**Review by ethanmills — COMMENTED** [review]
haven't made it all the way through yet, will come back to this in the morning

File: app/controllers/admin/participants_controller.rb (hunk 5 above)
Comment (kicferk1) [reply app/controllers/admin/participants_controller.rb:14]: Cause `school` is no longer an association on `ParticipantProfile`, so `ransack` does not work :(

File: app/controllers/admin/schools/participants_controller.rb (hunk 6 above)
Comment (kicferk1) [reply app/controllers/admin/schools/participants_controller.rb:13]: I doubt it, participants come from `school_cohorts` on the `school`. Which cannot have anything NPQ related.

File: app/controllers/schools/add_participants_controller.rb (hunk 7 above)
Comment (kicferk1) [reply app/controllers/schools/add_participants_controller.rb:14]: ehh, I really should call the scopes `ecf_participants` and `active_ecf_participants`. I hate NPQ profiles being participants.

File: app/forms/participant_mentor_form.rb (hunk 8 above)
Comment (kicferk1) [reply app/forms/participant_mentor_form.rb:16]: I don't know. Maybe. It feels like there are a bunch of tests missing.

Comment (kicferk1) [reply app/forms/participant_mentor_form.rb:16]: Turns out yes, it was. Good spot.

File: app/models/user.rb (hunk 12 above)
Comment (kicferk1) [reply app/models/user.rb:100]: When we move to `has_many` I imagine it will be a simple matter of switching the name to plural, right?

Am I missing something really scary?

Comment (ethanmills) [reply app/models/user.rb:99]: There are some implicit assumptions that there will only be one, I think

Comment (ethanmills) [reply app/models/user.rb:99]: Especially around display of NPQ/mentors

Comment (kicferk1) [reply app/models/user.rb:99]: I think the only place where we surface npqs and mentors at the same time is the admin panel, right?

I don't really know how we want to display them there, the obvious minimal implementation would display two copies of the same user, right?

Any more sophisticated solution is a bit out of my comfort zone and we should probably get a designers' opinion. Like, we could list the roles for each user, we could have a separate page for searching npq participants as opposed to ecf participants, we could probably have a few more ways of solving this problem. And I am not sure we know which one would be the most useful.

Comment (ethanmills) [reply app/models/user.rb:99]: No action here, just expressing fear

**Review by ethanmills — APPROVED** [review]

**Comment by ethanmills** [comment]
Looks like you've change the sort order on the participant listing page. it should be date added
